File: Day1-9/5.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINES 2048

int part1(int *jumps, int len) {
    int c = 0;
    int steps = 0;
    while(c >= 0 && c < len) {
        int j = jumps[c];
        jumps[c]++;
        steps++;
        c += j;
    }
    return steps;
}

int part2(int *jumps, int len) {
    int c = 0;
    int steps = 0;
    while(c >= 0 && c < len) {
        int j = jumps[c];
        if(j >= 3) {
            jumps[c]--;
        }
        else {
            jumps[c]++;
        }
        steps++;
        c += j;
    }
    return steps;
}

int getInput(char *fname, int** arrOut) {
    FILE *fp = fopen(fname, "r");
    char *line = NULL;
    size_t len = 0;
    int *arr = malloc(sizeof(int) * MAX_LINES);
    int i = 0;
    while (getline(&line, &len, fp) != -1) {
       arr[i++] = atoi(line);
    }
    free(line);
    *arrOut = arr;
    return i;
}

int main() {
    int *input = NULL;
    int len = getInput("5.txt", &input);

    int *p1 = malloc(sizeof(int) * len);
    memcpy(p1, input, sizeof(int) * len);
    int p1_result = part1(p1, len);
    printf("Part 1: %d\n", p1_result);
    
    free(p1);
    p1 = NULL;

    int *p2 = malloc(sizeof(int) * len);
    memcpy(p2, input, sizeof(int) * len);
    int p2_result = part2(p2, len);
    printf("Part 2: %d\n", p2_result);
    
    free(p2);
    p2 = NULL;

    free(input);
    input = NULL;
}

File: Day1-9/5.py
def first_jump_outside(jumps, mutator):
    c = 0
    steps = 0
    while c >= 0 and c < len(jumps):
        j = jumps[c]
        jumps[c] = mutator(j)
        steps += 1
        c += j
    return steps

if __name__ == '__main__':
    with open('5.txt') as f:
        jumps = [int(i) for i in f.readlines()]
    part1 = (lambda x: x+1)
    part2 = (lambda x: x-1 if x >= 3 else x+1)
    # example = [0, 3, 0, 1, -3]
    # assert first_jump_outside(example[:], part1) == 5
    # assert first_jump_outside(example[:], part2) == 10
    print("Part 1: {}".format(
        first_jump_outside(jumps[:], lambda x: x+1)
    ))
    print("Part 2: {}".format(
        first_jump_outside(jumps[:], lambda x: x-1 if x >= 3 else x+1)
    ))
